### ISSUE TITLE
how to run your code

### ISSUE THREAD

**Issue by lmlyzxiaop** [takuo-h/GNN-for-OOKB#1]
In your code ,the dataset is FB13,I replace the default dataset with dataset/OOKB/both-1000，train,dev,aux,test.And then I run the code ,I get (no such model H) ,could you help me?

**Comment by takuo-h** [comment]
Sorry, it's my mistake. I fixed it (maybe).  Because now I'm writing papers including phd thesis, I'm too busy to check it. However,  I'll check it at the latest within this year. thank you!

**Comment by takuo-h** [comment]
( maybe I need to adjust my codes because I use neural networks library, chainer, which has been updated recently )

**Comment by lmlyzxiaop** [comment]
Thanks for your reply,there are also some problems in your codes.I will try to learn chainer and fix your codes.

**Comment by lmlyzxiaop** [comment]
I have a problem.Is the result write in the margin file when training the model?

**Comment by lmlyzxiaop** [comment]
And where is the classification result?

**Comment by takuo-h** [comment]
For the first question: we calculate scores (margins) only contained in dev and test.  
For the second question: The evaluate function in main.py saves the scores (margins). To get result of triplet classification, you need to decide a threshold to classify whether dev/test triplets related with the scores are correct or not.

**Comment by takuo-h** [comment]
I'll try to run this code tonight.

**Comment by takuo-h** [comment]
(sorry for may late response)

**Comment by lmlyzxiaop** [comment]
I have already run your code in my machine,there are two problems in your code.First in the backend.py in array_int and array_float the 'volatile' is not allowed to be used anymore.So I just delete the volatile.Second,in modelI.py in  __call__ the dropout is not surpport 'train' ，I relace it with 'with chainer.using_config('train',is_train):    x = F.dropout(x,ratio=self.dropout_rate).

**Comment by takuo-h** [comment]
The fixs are right, because both problems arise from update of chainer.   
In particular, remove volatile in backend, and train in dropout to prevent the errors. 
And if you want to realize them (e.g., for reducing memories),  add the following code to dev/test case. 
with chainer.using_config('train',False), chainer.no_backprop_mode():

**Comment by lmlyzxiaop** [comment]
Thanks for your suggestions,I will try to add the code.

**Comment by takuo-h** [comment]
Maybe I did it. If you download this repository and unzip each datasets, you can run this code by "python main.py" only. (I hope so)

**Comment by takuo-h** [comment]
If you want to use new your model, write the model in "models" dir and add the model's name in manager.py in "models" dir. And then select your model using options '-nn your model name'. For example, the code "python main.py -nn X " run model X.

**Comment by lmlyzxiaop** [comment]
Okay, thank you very much. That's awesome

**Comment by takuo-h** [comment]
It is a pleasure, asking me! But this update is a rush job. So if you find bugs or errors you cannot solve, the please report it. Thank you!

**Comment by lmlyzxiaop** [comment]
In the margin file,the form is <head,relation,tail,label,score>。Should  I use the dev results to find the best threshold and then use it in test to get the classification result?

**Comment by takuo-h** [comment]
Yes. In general, dev dataset is used for finding best parameters including model parameters, architecture, and other hyper-parameters. And this case is too.

**Comment by lmlyzxiaop** [comment]
Okay,I got it, I thought you implement the classification function in your codes,but I did not find it. Thank you very much.

**Comment by takuo-h** [comment]
I added an code drawing score-history and related examples in draw-score-history dir. You can use the code to check how well your models learn.   
I did binary classification using threshold. In particular,  if a score is lower than the threshold, corresponding triplet is correct, and  vice versa. These procedures are needed, because our model use previous score functions, which provides scores only. And modifying and improving the score functions are not our interest, so our interest is to deal with OOKB.

**Comment by lmlyzxiaop** [comment]
Thank you very much to take time answer my questions and adjust your codes.I will read the codes tomorrow.Good night

**Comment by takuo-h** [comment]
Me too. I'll go to the bed soon after finishing my tasks... Good night!

**Comment by takuo-h** [comment]
I refined the drawing script and added an example script getting thresholds from scores. If you want to use other methods to get thresholds, add methods in the script.

**Comment by takuo-h** [comment]
Then, (maybe) I focus my papers. Thank you!

**Comment by lmlyzxiaop** [comment]
Thank you very much.

**Comment by lmlyzxiaop** [comment]
I have a problem in using gpu to run your model. I am sure my gpu is fine and I use  python main.py -g ，the speed is no difference from using cpu,that too slow.

**Comment by takuo-h** [comment]
To accelerate computation using gpu, we need to use matrix vector calculation as much as possible. In other words, we need to get together calculation in models. But, it is difficult for GNN to realize it, at least I can't. While I don't know how slow it is,  using gpu does not accelerate this implementations so much.

**Comment by lmlyzxiaop** [comment]
I got it . Thanks for all the things you have done. Thank you!

**Comment by lmlyzxiaop** [comment]
I find a problem in your dataset ,there is an entity that shown in the test of WN11 but not shown in train

**Comment by takuo-h** [comment]
Removing them is the easiest way. But if you want use them, initialize corresponding vectors. As a matter of fact, they do not effect evaluation because such entities are not so much contained. (if I remember correctly)

**Comment by lmlyzxiaop** [comment]
OK,I'll try.Thank you!
